### ISSUE TITLE
Creating Gordon Surfaces

### ISSUE THREAD

**Issue by gumyr** [gumyr/build123d#831]
### Discussed in https://github.com/gumyr/build123d/discussions/830

<div type='discussions-op-text'>

<sup>Originally posted by **jsbroks** December 19, 2024</sup>
I'm wondering what is the best way to create a Gordon Surface?

Here is what I have so far but as you would expect I getting `Error building a non-planer face with the provided exterior`.

```python
import csv
import numpy as np
from build123d import *


def read_heightmap(filename: str):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        heightmap = []
        for row in reader:
            heightmap.append([float(val) for val in row])
    return np.array(heightmap)


heightmap = read_heightmap('heightmap.csv')

points = []
for x in range(heightmap.shape[0]):
    for y in range(heightmap.shape[1]):
        points.append(Vector(x, y, heightmap[x,y]))


x_splines = []
y_splines = []


for x in range(heightmap.shape[0]):
    points = [(x, y, heightmap[x,y]) for y in range(heightmap.shape[1])]
    x_splines.append(Spline(points))


for y in range(heightmap.shape[1]):
    points = [(x, y, heightmap[x,y]) for x in range(heightmap.shape[0])]
    y_splines.append(Spline(points))

boundary_wire = Wire.make_wire([x_splines[0], y_splines[0], x_splines[-1], y_splines[-1]])

show_object(Face.make_surface(boundary_wire))
```</div>

**Comment by gumyr** [comment]
From @mnesarco

TiGL[1] has an OpenCascade based Gordon surface builder that is very good (c++). Christophe Grellier (@tomate44) ported it to FreeCAD's Curves WB[2] long time ago in Python. It can be used as a reference to port it to build123d. I am in the process of reorganizing the implementation as a standalone python module, but it is WIP and i can't make compromises on that [3]

[1] https://dlr-sc.github.io/tigl/
[2] https://github.com/tomate44/CurvesWB/blob/master/freecad/Curves/gordon.py
[3] https://github.com/mnesarco/MarzWorkbench/blob/master/freecad/marz/curves/gordon.py

**Comment by gongfan99** [comment]
I just wanted to share that I have ported rainman110's occ_gordon (https://github.com/rainman110/occ_gordon) to OCP line by line.
@gumyr I am thinking of releasing it to a individual python package `ocp_gordon` and then add a classmethod to the `Face` class. Are you open to add a new python package to build123d?
```
class Face:
    """
    Extends build123d.Face with a gordon_surface method for interpolation.
    """

    @classmethod
    def gordon_surface(
        cls,
        profiles: Union[List[Edge], ShapeList[Edge]],
        guides: Union[List[Edge], ShapeList[Edge]],
        tolerance: float = 1e-4,
    ) -> Face:
        """
        Creates a Gordon surface from a network of profile and guide curves.

        Args:
            profiles: A list or ShapeList of build123d.Edge objects representing profile curves.
            guides: A list or ShapeList of build123d.Edge objects representing guide curves.
            tolerance: Tolerance for surface creation and intersection calculations.

        Returns:
            A Face object representing the interpolated Gordon surface.
        """
```
BTW, it is a quite powerful tool. With only 2 profiles (i.e. 2 circles at one end) and 4 guides, an aircraft engine shell can be built:

<img width="745" height="208" alt="Image" src="https://github.com/user-attachments/assets/2584dd35-4611-4e45-9dbf-fb28b437dc36" />

**Comment by gumyr** [comment]
Yes, I would very much like to have Gordon surfaces as part of build123d - it's one of the items in the Roadmap (https://github.com/gumyr/build123d/wiki/Roadmap#gordon-surface).  Your Python package would need to be hosted on pypi as pypi doesn't accept GitHub packages. The svg importer that build123d uses is done this way if a reference is helpful: https://pypi.org/project/ocpsvg/0.3.4/.

Thanks for doing this!

**Comment by jdegenstein** [comment]
@gongfan99 This is great news! Is the repository public? Is it a pure python package?

**Comment by gongfan99** [comment]
@jdegenstein Yes, it is pure python. I just converted the original C++ code to python code strictly line by line, with the help of AI of course 🙂 
I will clean up the code and release it to pypi and github soon.

**Comment by gongfan99** [comment]
FYI... I have published the package at https://pypi.org/project/ocp-gordon
The source code is at https://github.com/gongfan99/ocp_gordon

Feel free to let me know if there is any issue.

Next I will add gordon_surface() to build123d.Face class.

**Comment by gongfan99** [comment]
@gumyr I have created a pull request to add the gordon surface. I have followed everything in the CONTRIBUTING.md

The only error is a type error:
src\build123d\topology\two_d.py:887: error: Argument 1 to "Curve_s" of "BRep_Tool" has incompatible type "Any | None"; expected "TopoDS_Edge"  [arg-type]
src\build123d\topology\two_d.py:888: error: Argument 1 to "Curve_s" of "BRep_Tool" has incompatible type "Any | None"; expected "TopoDS_Edge"  [arg-type]

It appears that this is a common error because Edge.wrapped has the "Unknown | None" type.

**Comment by gumyr** [comment]
Complete with commit 0ca0598907c02c3f9eebaa94d3bc60a296914357
